ttnn/cpp/ttnn/operations/ccl/kernel_common/worker_sync_utils.hpp

```diff
@@ -0,0 +1,231 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "dataflow_api.h"
+#include "debug/assert.h"
+#include "debug/dprint.h"
+#include "ttnn/cpp/ttnn/operations/ccl/shared_with_host/hetergeneous_data_structs.hpp"
+
+// Called by the master worker to synchronize with the slave workers
+FORCE_INLINE void master_sync_slaves(
+    const uint32_t num_workers_to_sync,
+
+    /* Used to get slave worker's sem addrs */
+    const uint32_t* worker_noc_coords,
+    const uint32_t worker_sync_sem_addr,
+
+    const uint64_t remote_op_l1_semaphore_addr) {
+
+    // Wait for all the slaves to finish their work
+    volatile tt_l1_ptr uint32_t* master_l1_semaphore_addr = reinterpret_cast<volatile tt_l1_ptr uint32_t*>(worker_sync_sem_addr);
+    noc_semaphore_wait(master_l1_semaphore_addr,  num_workers_to_sync - 1);
+    DPRINT << "MASTER SYNCED WITH SLAVES" << ENDL();
+
+    // Send signal to op
+    noc_semaphore_inc(remote_op_l1_semaphore_addr, 1);
+    DPRINT << "MASTER SIGNALED REMOTE OP" << ENDL();
+
+    // Clear the master semaphore, so that it can be used again
+    noc_semaphore_set(master_l1_semaphore_addr, 0);
+
+    // Clear the slave semaphores, so that they can continue processing
+    for (uint32_t i = 1; i < num_workers_to_sync; i++) { // Skip the first set of coords because they are for master worker
+        uint64_t remote_slave_l1_sem_addr = get_noc_addr(worker_noc_coords[i * 2], worker_noc_coords[i * 2 + 1], worker_sync_sem_addr);
+        noc_semaphore_inc(remote_slave_l1_sem_addr, 1);
+        DPRINT << "MASTER CLEAREED A SLAVE SEMAPHORE" << ENDL();
+    }
+}
+
+
+// Called by the slave worker to synchronize with the master worker
+FORCE_INLINE void slave_sync_master(
+    const uint32_t* worker_noc_coords,
+    const uint32_t worker_sync_sem_addr) {
+
+    // Signal the master that the slave has finished its work
+    uint64_t remote_master_l1_semaphore_addr = get_noc_addr(worker_noc_coords[0], worker_noc_coords[1], worker_sync_sem_addr);
+    noc_semaphore_inc(remote_master_l1_semaphore_addr, 1);
+    DPRINT << "SLAVE SYNCED WITH MASTER" << ENDL();
+
+    // Wait for the master to signal that this slave is ready to continue
+    volatile tt_l1_ptr uint32_t* slave_l1_semaphore_addr = reinterpret_cast<volatile tt_l1_ptr uint32_t*>(worker_sync_sem_addr);
+    noc_semaphore_wait(slave_l1_semaphore_addr, 1);
+    DPRINT << "SLAVE SEMAPHORE CLEARED BY MASTER" << ENDL();
+
+    // Clear the slave semaphore, so that it can be used again
+    noc_semaphore_set(slave_l1_semaphore_addr, 0);
+}
+
+
+FORCE_INLINE bool is_master(uint32_t master_x, uint32_t master_y, uint32_t worker_x, uint32_t worker_y) {
+    return master_x == worker_x && master_y == worker_y;
+}
+
+uint32_t increment_arg_idx(uint32_t& arg_idx, uint32_t num_args=1) {
+    uint32_t old_arg_idx = arg_idx;
```

Comment:
why not just inline this as rt_args_idx++?

Reply:
It's needed when the index is incremented by `num_workers_to_sync * 2`. For the remaining, I've reverted to `rt_args_idx++`